Add SFTP support

wertha opened the pull request.

wertha committed:
Add SFTP support

wertha committed:
Update README

Thomas-Smyth commented on client/ftp.js:
```diff
@@ -0,0 +1,29 @@
+import Client from 'basic-ftp';
+
+export class FTPClient {
+  constructor(timeout, logger, encoding) {
```
Can we pass these as an object please, i.e.:
```js
constructor(options = {}) {
    this.client = new Client(options.timeout);
    this.client.ftp.encoding = options.encoding || this.client.ftp.encoding;
    this.client.ftp.log = options.logger;
}
```

Thomas-Smyth commented on client/index.js:
```diff
@@ -0,0 +1,4 @@
+import { FTPClient } from "./ftp.js";
+import { SFTPClient } from "./sftp.js";
+
+export { FTPClient, SFTPClient };
```
Can you make sure prettier is run against your code? It should have been done when you committed, but it can also be done via `yarn run lint`.

In this case `"` should be `'` and there should be a new line at the end of the file.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+import sftp from 'ssh2-sftp-client';
+
+export class SFTPClient {
+  constructor(timeout, logger, encoding) {
```
Same as before, can we pass these as an object, please.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  async size(path) {
+    if (this.connectionClosed) throw Exception('Connect first in order to use this function');
```
Whilst this error is correct, I don't think we should include it:
 * This error is not present in the `ftp` version, or other similar locations, so it's inconsistent.
 * This error can only occur if the user of the package calls methods we haven't exposed in the documentation and it's a given that if you call random methods then errors are going to be thrown. However, if a user did decide to do this for whatever reason then I think the sftp package's error messages are going to be more helpful to them than error messages we throw to hide those errors.

If this is removed, I think `connectionClosed` will become obsolete too so that can probably be removed - that will also simplify this portion of the code and make it more in line with the `ftp` version.

wertha replied:
connectionClosed cannot be removed since it is being used when the client tries to connect. Since I couldn't get a state of the connection I handle it with this variable.

Thomas-Smyth replied:
I think we can still remove this error though, which you've done - I'll resolve this conversation.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+  async size(path) {
+    if (this.connectionClosed) throw Exception('Connect first in order to use this function');
+
+    this.log(`Getting size for file ${path}`);
```
Could we remove this please, we already have this log message elsewhere:
https://github.com/Thomas-Smyth/ftp-tail/blob/3dd9a9ac1cbcce17985af3f536e7a562d0ea3964/index.js#L88

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+    options.username = options.user;
+    options.readyTimeout = this.timeout;
+
+    this.log('Setting up sftp connection');
```
Could we remove this please, we already have this log message elsewhere:
https://github.com/Thomas-Smyth/ftp-tail/blob/3dd9a9ac1cbcce17985af3f536e7a562d0ea3964/index.js#L161

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+
+    this.log(`Getting size for file ${path}`);
+
+    return this.connection
+      .then(() => {
+        return this.client.stat(path);
+      })
+      .then((data) => {
+        this.log(`File size ${data.size}`);
+        return data.size;
+      })
+      .catch((err) => {
+        console.error(err.message);
+      });
```
Why is `connection` a promise here? I think we should wait for that to resolve in the `access` method rather than waiting for it to resolve later.

As a side note, could we remove `.then/.catch` from here in favour of `async/await` - I'm not a fan of mixing the two.

wertha replied:
why I decided not to wait, is because I didn't want to tied the pattern. But maybe it makes sense for all the wrappers to just be synchronous and not use await in the caller.

https://github.com/Thomas-Smyth/ftp-tail/blob/3be20cfbe95707360236e585b0001e3d8c4181fb/index.js#L156

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+    options.readyTimeout = this.timeout;
+
+    this.log('Setting up sftp connection');
+    this.connection = this.client.connect(options);
```
As above, I think we should be using `await` here.

wertha replied:
I didn't want to fixed that pattern that it is going to be used. The implementation calls the awaits, this works with promises just to keep it also consistent with the ftp library

Thomas-Smyth replied:
The `await` used by the implementation is waiting for the function this line is in to be finished, but because that function doesn't use `await` on this line it means that that function returns before the connection has occurred. You definitely should use an `await` on this line.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+      });
+  }
+
+  async access(options = {}) {
```
Maybe we should rename this method in both ftp clients to `connect` - I think `basic-ftp` client's use of `access` instead of `connect` is confusing.

wertha replied:
Will do, I also thought it was confusing, but left it because it was there already.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+  async downloadTo(destination, fromRemotePath, startAt = 0) {
+    if (this.connectionClosed) throw Exception('Connect first in order to use this function');
+
+    this.log(`Reading file starting from ${startAt}`);
```
Could we remove this please, we already have this log message elsewhere:
https://github.com/Thomas-Smyth/ftp-tail/blob/3dd9a9ac1cbcce17985af3f536e7a562d0ea3964/index.js#L107

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+    this.log(`Reading file starting from ${startAt}`);
+    this.readOptions.readStreamOptions.start = startAt;
+
+    return this.connection
+      .then(() => {
+        return this.client.get(fromRemotePath, destination, this.readOptions);
+      })
+      .catch((err) => {
+        console.error(err.message);
+      });
```
As above.

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  close() {
+    this.log('Closing connection');
```
Could we remove this please, we already have this log message elsewhere:
https://github.com/Thomas-Smyth/ftp-tail/blob/3dd9a9ac1cbcce17985af3f536e7a562d0ea3964/index.js#L170

Thomas-Smyth commented on client/ftp.js:
```diff
@@ -0,0 +1,29 @@
+    this.client.close();
+  }
+
+  get closed() {
```
Is this method used anywhere?

wertha replied:
Is used on the implementation.

https://github.com/Thomas-Smyth/ftp-tail/blob/3be20cfbe95707360236e585b0001e3d8c4181fb/index.js#L152-L159

Thomas-Smyth commented on client/sftp.js:
```diff
@@ -0,0 +1,83 @@
+    this.client.end();
+  }
+
+  get closed() {
```
As above, is this method used anywhere?

wertha replied:
same as before, on the funtions, connect and disconnect that property is being checked.

Thomas-Smyth replied:
Would it not just make more sense to rename `connectionClosed` to `closed` and get rid of this getter function? Currently, you have two properties with very similar names that do exactly the same thing as one another.

Thomas-Smyth commented on index.js:
```diff
@@ -16,23 +16,28 @@ export default class FTPTail extends EventEmitter {
     } else {
       this.log = () => {};
-      this.client.ftp.log = () => {};
+    }
```
We could actually get `ssh2-sftp-client`'s logging if we wanted by passing in the `debug` parameter:
https://github.com/theophilusx/ssh2-sftp-client/blob/master/src/index.js#L36

Thomas-Smyth reviewed (changes requested):
Good works, thanks for making the OR - just a few minor things regarding consistency, promises and stylistic bits to address, please.

wertha committed:
Fix PR issues

Remove log messages from sftp
Rename access to connect
Use an object in the constructor params
Remove throwing exception when connection is not initialized

Thomas-Smyth commented:
I have closed everything that's been addressed, but I think there's still some confusion regarding the promises.

_Foreword: I'll reference you're `connect` function as such and the sftp client's connect function as `this.client.connect`..._

In your `connect` function, you're calling `this.client.connect`, but not `await`ing the result of this:
https://github.com/Thomas-Smyth/ftp-tail/pull/4/files#diff-f36caa7b3db486c887d3a5335edbba698925a240b555fa303017836c6f0ad76aR46

This means that your `connect` function returns a promise that is resolved when the function **starts** connecting to the sftp server rather than when it **finishes** connecting to the sftp server. Thus, anything that `await`s your `connect` function will continue when the function **starts** connecting to the sftp server rather than when it **finishes** connecting to the sftp server, which is not behaviour we want - we do not want to try tailing files before we finish connecting to the sftp server as it would not be possible.

Therefore, you should change the line linked above to simply:
```js
await this.client.connect(options);
```

Note, the removal of the `this.connection = ` is intentional as now you have `await`ed the promise here, you do not need to `await` it elsewhere, which you are currently doing via `this.connection.then` in various places, so you do not need to store it anywhere. This also means that you can also simplify the `await`s, via `.then`, used on `this.connection` to:
```js
// size method:
return (await this.client.size(path)).size;

// downloadTo method:
return this.client.get(fromRemotePath, destination, this.readOptions);
```

The methods should look very similar lengthwise to the ftp version, i.e. the `size` method can also be 1 line long and the `downloadTo` can be 2 lines long.

I'll be away the latter half of this week so I might not be able to fully merge this PR until I get back.

wertha committed:
Fix connection promise

wertha commented:
Any updates?

naxhh commented:
any update on this? I started the task to make squadjs work with sftp to realize the work was done a long time ago

(I'll be happy to take this over and finish it but what is actually missing to get it merged?)

Thomas-Smyth commented:
Closed in favor of:
https://github.com/Thomas-Smyth/ftp-tail/commit/ee757ee5d3eaf0942088b8ede602d1f57360a3b9

Thank you for the contribution regardless.
